File: sammy/230425_study/프로_92345_사라지는 발판.py
n,m=0,0
dx=[1,-1,0,0]
dy=[0,0,1,-1]
visit=[[0]*5 for _ in range(5)]

def outArea(x,y):
    return x < 0 or x >= n or y < 0 or y >= m

def dfs(board,cur_x,cur_y,op_x,op_y):
    if visit[cur_x][cur_y]: return 0
    canWin = 0

    for i in range(4):
        nx,ny=cur_x+dx[i],cur_y+dy[i]
        if outArea(nx,ny) or visit[nx][ny] or board[nx][ny] == 0: continue
        visit[cur_x][cur_y]=1 # 방문처리
        op_result = dfs(board,op_x,op_y,nx,ny) + 1
        visit[cur_x][cur_y]=0 # 방문 끝남 처리

        # 현재 저장된 값 패배인데 상대가 졌다고 하면 이기는 경우로 저장
        if canWin % 2 == 0 and op_result % 2 == 1 : canWin = op_result
        # 현재 저장된 값 패배인데 상대가 이겼다고 하면 최대한 늦게 지는 횟수 선택
        elif canWin % 2 == 0 and op_result % 2 == 0 : canWin = max(canWin,op_result)
        # 현재 저장된 값 승리인데 상대가 졌다고 하면 최대한 빨리 이기는 횟수 선택
        elif canWin % 2 == 1 and op_result % 2 == 1 : canWin = min(canWin,op_result)
        
    return canWin

def solution(board, aloc, bloc):
    global n,m
    n,m=len(board),len(board[0])
    return dfs(board,aloc[0],aloc[1],bloc[0],bloc[1])
    


print(solution(	[[1, 1, 1], [1, 1, 1], [1, 1, 1]], [1, 0], [1, 2]),5)
print(solution(	[[1, 1, 1], [1, 0, 1], [1, 1, 1]], [1, 0], [1, 2]),4)
print(solution(	[[1, 1, 1, 1, 1]], [0, 0], [0, 4]),4)
print(solution(	[[1]], [0, 0], [0, 0]),0)

# 항상 이기는 플레이어 B, 먼저 시작하는 플레이어 A

File: James/230502_study/프로_92345_사라지는발판.java
import java.util.*;

class Solution {
    static int n, m;
    static int[] dx = {-1, 1, 0, 0};
    static int[] dy = {0, 0, -1, 1};
    
    static class Result {
        boolean win;
        int moveCnt;
        
        public Result(boolean win, int moveCnt) {
            this.win = win;
            this.moveCnt = moveCnt;
        }
    }
    
    public Result dfs(int[][] board, int ax, int ay, int bx, int by, int aDepth, int bDepth) {
        boolean win = false;
        int winCnt = 5 * 5;
        int loseCnt = aDepth + bDepth;
        
        // A 움직임
        if(aDepth == bDepth && board[ax][ay] == 1) {
            for(int i = 0; i < 4; i++) {
                int nx = ax + dx[i];
                int ny = ay + dy[i];
                
                if(!isMovePossible(board, nx, ny))
                    continue;
                
                board[ax][ay] = 0;
                Result r = dfs(board, nx, ny, bx, by, aDepth + 1, bDepth);
                win |= !r.win;
                
                if(r.win)
                    loseCnt = Math.max(loseCnt, r.moveCnt);
                
                else
                    winCnt = Math.min(winCnt, r.moveCnt);
                
                board[ax][ay] = 1;
            }
        }
        
        // B 움직임
        else if(aDepth > bDepth && board[bx][by] == 1) {
            for(int i = 0; i < 4; i++) {
                int nx = bx + dx[i];
                int ny = by + dy[i];
                
                if(!isMovePossible(board, nx, ny))
                    continue;
                
                board[bx][by] = 0;
                Result r = dfs(board, ax, ay, nx, ny, aDepth, bDepth + 1);
                win |= !r.win;
                
                if(r.win)
                    loseCnt = Math.max(loseCnt, r.moveCnt);
                
                else
                    winCnt = Math.min(winCnt, r.moveCnt);
                
                board[bx][by] = 1;
            }
        }
        
        return new Result(win, win ? winCnt : loseCnt);
    }
    
    public boolean isMovePossible(int[][] board, int nx, int ny) {
        if(nx < 0 || ny < 0 || nx >= n || ny >= m || board[nx][ny] == 0) {
            return false;
        } else {
            return true;
        }
    }
    
    public int solution(int[][] board, int[] aloc, int[] bloc) {
        n = board.length;
        m = board[0].length;
        
        return dfs(board, aloc[0], aloc[1], bloc[0], bloc[1], 0, 0).moveCnt;        
    }
}
